cartographer/cloud/internal/local_trajectory_uploader.cc

```diff
@@ -151,13 +152,18 @@ void LocalTrajectoryUploader::Shutdown() {
 }
 
 void LocalTrajectoryUploader::TryRecovery() {
-  auto channel_state = client_channel_->GetState(true /* try_to_connect */);
+  auto channel_state = client_channel_->GetState(false /* try_to_connect */);
   if (channel_state != grpc_connectivity_state::GRPC_CHANNEL_READY) {
-    LOG(ERROR) << "Failed to re-connect to uplink prior to recovery attempt.";
-    return;
-  } else {
-    LOG(INFO) << "Uplink channel ready again, trying recovery.";
+    LOG(INFO) << "Trying to re-connect to uplink...";
+    std::chrono::system_clock::time_point deadline =
+        std::chrono::system_clock::now() +
+        std::chrono::seconds(kConnectionRecoveryTimeoutInSeconds);
+    if (!client_channel_->WaitForConnected(deadline)) {
```

Comment:
for simplicity, you can just do this if (drop the previous one) and use one timeout.

Reply:
> WaitForConnected also uses GetState(true) internally, see here.

Ok, then I don't mind. Thanks for the clarification.

Reply:
The "Trying to re-connect to uplink..." is quite useful though...